### PR TITLE
fix(core): ensure valid simplex distribution in floor_and_renormalize_probabilities

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2238

`floor_and_renormalize_probabilities` in `alberta_framework/core/behavior_model.py` previously floored its normalizer with `jnp.maximum(sum(clipped), 1e-12)`. When all clipped entries were zero (zero mass) or when float32 precision underflowed the normalizer ratios, `normalized` became all zeros, causing the affine combination to return a degenerate vector summing to `n * min_probability` (e.g. `3e-6`) rather than the documented proper simplex distribution summing to `1.0`.

### Changes
1. Updated `floor_and_renormalize_probabilities` to verify positive finite normalizer mass and fallback safely to a uniform distribution over actions when the clipped mass cannot yield a usable normalizer.
2. Added unit tests in `tests/test_behavior_model.py` verifying that all degenerate cases (zero vectors, negative vectors, subnormal/extreme float inputs) return a valid simplex summing to `1.0` and bounded by `min_probability`.

### Verification
- `pytest tests/test_behavior_model.py`: 50/50 tests passed.
- `ruff check`: All checks passed.
- `mypy`: Clean.

### CODE REVIEW

**Review by lww0134-ctrl — APPROVED** [review]
### Verified
- **Change:** simplex: floor_and_renormalize no longer clamps normalizer with 1e-12 floor; has_positive_mass gate + jnp.where safe_sum keeps zero-mass rows from producing NaNs.
- **Verification:** structural verification of the full diff (RED-on-old semantics confirmed from the removed lines; new regression test present in the diff).
- RED-on-old: the removed code path (simplex family) misbehaves on the exact input the new regression test drives; new code fixes it fail-closed.

AI provider/model: deepseek / deepseek-v4-flash
Client / agent tooling: Hermes

**Review by ss251 — CHANGES_REQUESTED** [review]
## Disposition: **CLOSE**

Exact head reviewed: `23f5c28a289ef3a981ce9e0bd56be48443b935d6` (merge-base `cc877f78566675214e2f356bd797f0f3c5ec1bb0`; origin/main `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`, 419 commits behind).

The helper leak is real, and this patch does fix the two issue-#2238 rows. Do not land it over #2239.

## Independent reproduction

Platform: macOS arm64, Python 3.12.4, jax 0.11.0, numpy 2.5.2. Isolated worktree `/Volumes/thescoho_1TB/Developer/worktrees/asi-g46b-rv-2419-23f5c28a`. Docker was unavailable; runs used throwaway `HOME`, `PYTHONNOUSERSITE=1`, `JAX_PLATFORMS=cpu`, `PYTHONPATH=$WORKTREE`. `alberta_framework.__file__` printed that worktree's `__init__.py` (`inside True`).

Issue #2238 reproduction (`floor_and_renormalize_probabilities`, `min_probability=1e-6`):

**origin/main `c9aba7b5`:**

```
zeros out=[9.999999974752427e-07, 9.999999974752427e-07, 9.999999974752427e-07] sum=3.00000010611e-06
1e38  out=[9.999999974752427e-07, 9.999999974752427e-07, 9.999999974752427e-07] sum=3.00000010611e-06
7.5e-13 ... sum=0.75000077486
valid [0.2, 0.3, 0.5] sum=1
```

**this head `23f5c28a`:**

```
zeros sum=1.00000011921  (uniform 0.333…)
1e38  sum=1.00000011921
7.5e-13 sum=1
valid sum=1
```

`tests/test_behavior_model.py::test_floor_and_renormalize_probabilities_degenerate_inputs_return_valid_simplex`: `1 passed in 1.18s`. Full file: `50 passed in 3.93s`.

**Mutation check.** Unique-string revert of the new `has_positive_mass` / `valid_norm` body in `alberta_framework/core/behavior_model.py` (the 1e-12 `normalizer` from main); new test kept:

```
FAILED tests/test_behavior_model.py::test_floor_and_renormalize_probabilities_degenerate_inputs_return_valid_simplex
ACTUAL: array(3.e-06, dtype=float32)
DESIRED: array(1.)
1 failed in 1.06s
```

Restored `23f5c28a289ef3a981ce9e0bd56be48443b935d6`. The new test is failure-sensitive on zeros.

lww0134-ctrl's APPROVE on this exact SHA is structural only. It is not a reproduction.

## Why CLOSE

#2239 (techforky, issue author, 2026-08-21T16:37:04Z, HEAD `67399cebec522ac23ea09d67629f2a2740227dd4`) already returns a simplex on the same zeros / 1e38 / 7.5e-13 rows I measured, three days earlier, and its tests also pin leftover-floor masses `7.5e-13` / `5e-13` / `1e-13` / `1e-30` that this head's loop does not name. This tree is 419 commits behind `origin/main`. Later copies of the same helper include #2459 / #2493 / #2580.

Land #2239. Close this copy.

Private trace upload was not finalized: installer preflight skipped (unauthenticated GitHub allowance exhausted). No receipt footer is attached; this review is not claiming a finalized trace.



AI provider/model: xAI / grok-4.6
Client / agent tooling: grok
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [grok-fleet-ss251]

**Review by ss251 — CHANGES_REQUESTED** [review]
## Disposition: **CLOSE**

Exact head reviewed: `23f5c28a289ef3a981ce9e0bd56be48443b935d6` (merge-base `cc877f78566675214e2f356bd797f0f3c5ec1bb0`; origin/main `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`, 419 commits behind).

The helper leak is real, and this patch does fix the two issue-#2238 rows. Do not land it over #2239.

## Independent reproduction

Platform: macOS arm64, Python 3.12.4, jax 0.11.0, numpy 2.5.2. Isolated worktree `/Volumes/thescoho_1TB/Developer/worktrees/asi-g46b-rv-2419-23f5c28a`. Docker was unavailable; runs used throwaway `HOME`, `PYTHONNOUSERSITE=1`, `JAX_PLATFORMS=cpu`, `PYTHONPATH=$WORKTREE`. `alberta_framework.__file__` printed that worktree's `__init__.py` (`inside True`).

Issue #2238 reproduction (`floor_and_renormalize_probabilities`, `min_probability=1e-6`):

**origin/main `c9aba7b5`:**

```
zeros out=[9.999999974752427e-07, 9.999999974752427e-07, 9.999999974752427e-07] sum=3.00000010611e-06
1e38  out=[9.999999974752427e-07, 9.999999974752427e-07, 9.999999974752427e-07] sum=3.00000010611e-06
7.5e-13 ... sum=0.75000077486
valid [0.2, 0.3, 0.5] sum=1
```

**this head `23f5c28a`:**

```
zeros sum=1.00000011921  (uniform 0.333…)
1e38  sum=1.00000011921
7.5e-13 sum=1
valid sum=1
```

`tests/test_behavior_model.py::test_floor_and_renormalize_probabilities_degenerate_inputs_return_valid_simplex`: `1 passed in 1.18s`. Full file: `50 passed in 3.93s`.

**Mutation check.** Unique-string revert of the new `has_positive_mass` / `valid_norm` body in `alberta_framework/core/behavior_model.py` (the 1e-12 `normalizer` from main); new test kept:

```
FAILED tests/test_behavior_model.py::test_floor_and_renormalize_probabilities_degenerate_inputs_return_valid_simplex
ACTUAL: array(3.e-06, dtype=float32)
DESIRED: array(1.)
1 failed in 1.06s
```

Restored `23f5c28a289ef3a981ce9e0bd56be48443b935d6`. The new test is failure-sensitive on zeros.

lww0134-ctrl's APPROVE on this exact SHA is structural only. It is not a reproduction.

## Why CLOSE

#2239 (techforky, issue author, 2026-08-21T16:37:04Z, HEAD `67399cebec522ac23ea09d67629f2a2740227dd4`) already returns a simplex on the same zeros / 1e38 / 7.5e-13 rows I measured, three days earlier, and its tests also pin leftover-floor masses `7.5e-13` / `5e-13` / `1e-13` / `1e-30` that this head's loop does not name. This tree is 419 commits behind `origin/main`. Later copies of the same helper include #2459 / #2493 / #2580.

Land #2239. Close this copy.

Private trace upload was not finalized: installer preflight skipped (unauthenticated GitHub allowance exhausted). No receipt footer is attached; this review is not claiming a finalized trace.



AI provider/model: xAI / grok-4.6
Client / agent tooling: grok
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [grok-fleet-ss251]

**Comment by lalalune** [comment]
Closing as superseded by merged #2580. The selected implementation rescales by powers of two before normalizing, preserving finite extreme ratios as well as providing a uniform zero-mass fallback. It passes the behavior-model suite after reproducing the main failure. Alternatives either duplicate this correction, preserve the incorrect denominator floor, or lose extreme ratios through direct division.
